### PR TITLE
Add EKS Cluster Processor and Kubernetes Catalog Collector

### PR DESCRIPTION
**Pull request by backjo**
Porting this functionality from our internal backstage. 

There are two major pieces of functionality that we are introducing here. 
1) A new processor in the Catalog AWS plugin to ingest EKS clusters into the Catalog.
2) A new locator in the kubernetes-backend plugin that allows sourcing Kubernetes cluster information from the catalog. 

Tests to follow on this PR, but we'd like some initial validation on the approach and usefulness here before we land those. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [X] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by backjo** [comment]
Hi @jhaals !

While I agree that generally infrastructure should be fairly ephemeral, I think there's probably a case to be made for ingesting some level of infrastructure into the catalog. The fact that the existing Kubernetes plugin is configured out of the box with static cluster information points to K8S clusters being generally long-lived pieces of infrastructure. Also - I believe the existing S3 and Cloud Account processors fall into the same category as this processor - both can also be long lived or more ephemeral as well, depending on the use case and user. 

Just to add an additional use case - we leverage this catalog data heavily internally to provide Scaffolder actions around provisioning kubernetes namespaces and other resources, not just to enable the Kubernetes plugin. 

Happy to close this and manage our own plugin / keep it internal if it's not directionally aligned with the Catalog, but I do feel that having this in the core would provide benefit to end users. At the end of the day, I'd expect it to be up to the end user to decide if it makes sense to enable this processor and ingest this information.

**Review by jhaals — COMMENTED** [review]
Thanks for the clarification @backjo and sorry to the slow reply on this one.
I can see your use case and that the clusters are at a similar level of stability as for example an S3 bucket. @mclarke47 does this looks good for you for you in the Kubernetes plugin? In that case I think the next steps would be tests, changesets and possibly some docs for users :)

**Comment by mclarke47** [comment]
I really like the idea of having clusters as entities in the catalog. This opens up a lot of possibilities of where we could go (k8s cluster management through Backstage).

This might also be a replacement for the `clusterLocator` interface in the kubernetes plugin as maybe we could just rely on the catalog processors to define the cluster configuration.

**Comment by backjo** [comment]
Sweet! Let me go ahead and tidy up the PR with the docs/tests/etc so we're closer to a reviewable/merge ready state.

**Comment by backjo** [comment]
@jhaals @mclarke47 tidied this up a bit - would appreciate a review when you have some time!

File: docs/features/kubernetes/configuration.md [hunk 1]
```diff
@@ -57,12 +57,17 @@ This is an array used to determine where to retrieve cluster configuration from.
...
 Valid cluster locator methods are:
 
+- [`catalog`](#catalog)
 - [`config`](#config)
 - [`gke`](#gke)
 - [custom `KubernetesClustersSupplier`](#custom-kubernetesclusterssupplier)
 
 #### `config`
 
+This cluster locator method will read cluster information from the catalog.
```
Comment (freben) [inline docs/features/kubernetes/configuration.md:67]: now need to rename the heading above :)

File: plugins/catalog-backend-module-aws/src/processors/AwsEKSClusterProcessor.test.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AwsEKSClusterProcessor } from './AwsEKSClusterProcessor';
+import AWSMock from 'aws-sdk-mock';
+import aws from 'aws-sdk';
+
+describe('AwsOrganizationCloudAccountProcessor', () => {
```
Comment (freben) [inline plugins/catalog-backend-module-aws/src/processors/AwsEKSClusterProcessor.test.ts:21]: AwsEKSClusterProcessor

File: .changeset/plenty-books-peel.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/plugin-catalog-backend-module-aws': minor
+'@backstage/plugin-kubernetes-backend': minor
```
Comment (freben) [inline .changeset/plenty-books-peel.md:3]: The backend change is a minor since it's breaking, and could be in its own changeset. It could contain something that helps users apply the right changes to their own instance to follow along when upgrading. For example:

> The `KubernetesBuilder.createBuilder` method now requires an additional field,
> `discovery`. To update your backend, you will want to do something like the following:
>
> ```diff
>  KubernetesBuilder.createBuilder({
>    config: env.config,
>    logger: env.config,
> +  discovery: env.discovery,
>  })
> ```

File: plugins/catalog-backend-module-aws/src/processors/AwsEKSClusterProcessor.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import { AWSCredentialFactory } from '../types';
+
+const ACCOUNTID_ANNOTATION: string = 'amazonaws.com/account-id';
+const ARN_ANNOTATION: string = 'amazonaws.com/arn';
+const KUBERNETES_API_SERVER_ANNOTATION = 'kubernetes.io/api-server';
+const KUBERNETES_API_SERVER_CA_ANNOTATION =
+  'kubernetes.io/api-server-certificate-authority';
+const KUBERNETES_AUTH_METHOD_ANNOTATION = 'kubernetes.io/auth-provider';
+
+export class AwsEKSClusterProcessor implements CatalogProcessor {
```
Comment (freben) [inline plugins/catalog-backend-module-aws/src/processors/AwsEKSClusterProcessor.ts:31]: I don't see any api-report change for this package - did you forget to export it all the way?

**Review by freben — COMMENTED** [review]
Just a quick initial look to begin with.

Could this have been an entity provider instead of a processor? That does have benefits, and is the main method of ingesting data these days.

https://backstage.io/docs/features/software-catalog/external-integrations

File: plugins/catalog-backend-module-aws/src/processors/AwsEKSClusterProcessor.test.ts (hunk 2 above)
Comment (backjo) [reply plugins/catalog-backend-module-aws/src/processors/AwsEKSClusterProcessor.test.ts:21]: yep

**Comment by backjo** [comment]
> Just a quick initial look to begin with.
> 
> Could this have been an entity provider instead of a processor? That does have benefits, and is the main method of ingesting data these days.
> 
> https://backstage.io/docs/features/software-catalog/external-integrations

@freben will rely on your expertise here - the reason we chose to go with a processor is that we extend the AwsCloudOrganizationProcessor internally to emit locations for every cloud account - which then gives this processor the information it needs to go and fetch potential entities from those accounts.

File: plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    this.catalogClient = catalogClient;
+  }
+
+  static fromConfig(catalogApi: CatalogApi): CatalogClusterLocator {
+    return new CatalogClusterLocator(catalogApi);
+  }
+
+  async getClusters(): Promise<ClusterDetails[]> {
+    const clusters = await this.catalogClient.getEntities({
+      filter: [
```
Comment (freben) [inline plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.ts:33]: Filters effectively have an OR between array entries. In its current form, this will return items that EITHER have `spec.type: kubernetes-cluster`, OR annotation a, etc. If I understand your logic here, this should be an array with one single object, that has four keys in it.

File: plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        { 'spec.type': 'kubernetes-cluster' },
+        {
+          'metadata.annotations.kubernetes.io/api-server':
```
Comment (freben) [inline plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.ts:36]: You should probably import the constants, e.g. KUBERNETES_API_SERVER_CA_ANNOTATION etc and use those here.

File: plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        {
+          'metadata.annotations.kubernetes.io/auth-provider':
+            CATALOG_FILTER_EXISTS,
+        },
+      ],
+    });
+    return clusters.items.map(entity => {
+      const clusterDetails: ClusterDetails = {
+        name: entity.metadata.name,
+        // @ts-ignore filtered out by catalog-client query.
```
Comment (freben) [inline plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.ts:52]: Avoid ts-ignore like the plague :) You can place an exclamation mark on the line below instead, e.g.

```ts
url: entity.metadata.annotations![KUBERNETES_API_SERVER_ANNOTATION]!
```

File: plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.ts (hunk 5 above)
Comment (freben) [reply plugins/kubernetes-backend/src/cluster-locator/CatalogClusterLocator.ts:33]: Come to think of it, shouldn't you also check `kind: Resource`?

**Review by freben — COMMENTED** [review]
Almost done!

File: plugins/kubernetes-backend/src/cluster-locator/index.ts [hunk 8]
```diff
@@ -38,13 +41,17 @@ class CombinedClustersSupplier implements KubernetesClustersSupplier {
 
 export const getCombinedClusterSupplier = (
   rootConfig: Config,
+  discovery: PluginEndpointDiscovery,
   refreshInterval: Duration | undefined = undefined,
 ): KubernetesClustersSupplier => {
+  const catalogClient = new CatalogClient({ discoveryApi: discovery });
```
Comment (mclarke47) [inline plugins/kubernetes-backend/src/cluster-locator/index.ts:47]: If we just use discovery to create a CatalogClient, why don't we just pass the catalog client we already create from discovery [here]([standaloneApplication.ts](https://github.com/backstage/backstage/pull/11789/files#diff-d343f848736906b291d4a16a7bbc66b327fccd20b46bbffaa558e404c670f09e))

**Comment by mclarke47** [comment]
LGTM Just one comment to discuss :)

File: plugins/kubernetes-backend/src/cluster-locator/index.ts (hunk 8 above)
Comment (freben) [reply plugins/kubernetes-backend/src/cluster-locator/index.ts:47]: 👍 and we also already pass a client into the original builder, so should be available in all cases now. Maybe the added discovery parameter can be removed everywhere then

File: plugins/kubernetes-backend/package.json [hunk 9]
```diff
@@ -40,6 +40,7 @@
     "@backstage/catalog-client": "^1.0.4-next.1",
     "@backstage/catalog-model": "^1.1.0-next.2",
     "@backstage/config": "^1.0.1",
+    "@backstage/core-plugin-api": "^1.0.3-next.0",
```
Comment (freben) [inline plugins/kubernetes-backend/package.json:43]: Oh this should not be here - this would be a dependency from a backend plugin to a frontend core.

**Review by freben — COMMENTED** [review]
Almost there!

**Comment by backjo** [comment]
> @backjo sorry a small conflict has popped itself in here, could you fix it up? pray

Addressed!

**Review by benjdlambert — APPROVED** [review]
OK! Let's get this in an iterate! :ship:

**Comment by asaf-erlich** [comment]
@backjo or anyone else who was part of getting this PR merged, can someone provide an example of how the app config needs to look and the `packages/backend/src/plugins/catalog.ts` needs to look to get this working? I would love to get this working but am having trouble trying to figure it out since there is zero documentation or online open source examples to look at.

The closing thing I suspect to it is the s3 discovery / AwsS3EntityProvider which is also [part of the same plugin](https://backstage.io/docs/reference/plugin-catalog-backend-module-aws). That one has documentation https://backstage.io/docs/integrations/aws-s3/discovery and https://backstage.io/docs/integrations/aws-s3/locations but the interface is different for [AwsEKSClusterProcessor](https://backstage.io/docs/reference/plugin-catalog-backend-module-aws.awseksclusterprocessor) in the code.

Thank you and very much appreciate this contribution,
-Asaf

**Comment by asaf-erlich** [comment]
Created a github issue with a problem I had trying to get this working: https://github.com/backstage/backstage/issues/13191 . It does contain some example configuration details. If I get help getting that resolved and working I can also contribute the missing documentation.

**Comment by mhamrah** [comment]
I came across this trying to get Kubernetes data into backstage. If I understand this functionality correctly, it simply uses aws-eks as a location store for catalog-info.yaml files? I added a location type and target to my app-config.yaml, but don't see anything in the UI relating to autodiscovery of clusters, etc. Hoping to understand what the purpose of this is- it doesn't seem to show kubernetes cluster for the kubernetes integration of services.

**Comment by asaf-erlich** [comment]
@mhamrah I think the clusters show up in the catalog as type resource. In my local fork I added logging since there's no indication it works if you don't know where to look. If you look at my linked github issue above it includes a working configuration example.

**Comment by lukasz-bielinski** [comment]
Hi @backjo , 

could you please point me, how to configure this:
1. I have autodiscovered eks clusters
2. every cluster has dedicated role to access k8s(logs, list of pods etc)

So in backstage I would like to have `main` role, which can assume roles from another clusters. Currently I've configured that role from 1st cluster can assume role from the second.

right now I have:

```
aws:
  mainAccount:
    roleName: ${MAIN_ROLE_AWS_ARN}
    region: ${AWS_REGION}
    accounts:
      -  roleName: ${DOWNSTREAM_ROLE_AWS_ARN}
          region: ${AWS_REGION}
```
but it does not work. Any help highly appreciated :)

backstage-assume-role
on the screenshot:
1. first cluster - info about pods is avaiable
2. second - UNAUTHORIZED_ERROR when accessing info about pods
